Development/nmos/ocsp_behaviour.cpp

```diff
@@ -0,0 +1,378 @@
+    }
+
+    void ocsp_behaviour_thread(nmos::model& model, nmos::experimental::ocsp_settings& ocsp_settings, load_ca_certificates_handler load_ca_certificates, load_server_certificates_handler load_server_certificate, slog::base_gate& gate_)
+    {
+        nmos::details::omanip_gate gate(gate_, nmos::stash_category(nmos::categories::ocsp_behaviour));
+
+        details::ocsp_behaviour_thread(model, ocsp_settings, load_ca_certificates, load_server_certificate, gate);
+    }
+
+    // callbacks from this function are called with the model locked, and may read or write directly to the model
+    void details::ocsp_behaviour_thread(nmos::model& model, nmos::experimental::ocsp_settings& ocsp_settings, load_ca_certificates_handler load_ca_certificates, load_server_certificates_handler load_server_certificate, slog::base_gate& gate)
+    {
+        enum
+        {
+            initial_ocsp_operation,
+            ocsp_behaviour
+        } mode = initial_ocsp_operation;
+
+        std::vector<web::uri> ocsp_uris;
+
+        nmos::details::seed_generator backoff_seeder;
+        std::default_random_engine backoff_engine(backoff_seeder);
+        double backoff = 0;
+
+        ocsp_shared_state state{ load_ca_certificates };
+
+        // continue until the server is being shut down
+        for (;;)
+        {
+            if (with_read_lock(model.mutex, [&] { return model.shutdown; })) break;
+
+            switch (mode)
+            {
+                case initial_ocsp_operation:
+                {
+                    // Note: use the same discovery backoff settings for OCSP API retry
+                    if (0 != backoff)
+                    {
+                        auto lock = model.read_lock();
+                        const auto random_backoff = std::uniform_real_distribution<>(0, backoff)(backoff_engine);
+                        slog::log<slog::severities::more_info>(gate, SLOG_FLF) << "Waiting to retry OCSP API for about " << std::fixed << std::setprecision(3) << random_backoff << " seconds (current backoff limit: " << backoff << " seconds)";
+                        model.wait_for(lock, std::chrono::milliseconds(std::chrono::milliseconds::rep(1000 * random_backoff)), [&] { return model.shutdown; });
+                        if (model.shutdown) break;
+                    }
+
+                    // get the list of server certificates chain data
+                    const auto server_certificates = load_server_certificate();
+                    const auto server_certificate_chains = boost::copy_range<std::vector<utility::string_t>>(server_certificates | boost::adaptors::transformed([](const nmos::certificate& cert) { return cert.certificate_chain; }));
+
+                    try
+                    {
+                        // get OCSP URIs
+                        ocsp_uris = details::get_ocsp_uris(server_certificate_chains, gate);
```

Comment:
Possible bug 🐛 If `get_ocsp_uris` throws, will this thread spin in the `initial_ocsp_operation` state without increasing the `backoff`?